src/loaders/api/register-api.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright 2020-2025 New Relic, Inc. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { handle } from '../../common/event-emitter/handle'
+import { warn } from '../../common/util/console'
+import { isValidTarget } from '../../common/util/target'
+import { FEATURE_NAMES } from '../features/features'
+import { now } from '../../common/timing/now'
+import { SUPPORTABILITY_METRIC_CHANNEL } from '../../features/metrics/constants'
+
+/**
+ * @typedef {import('./register-api-types').RegisterAPI} RegisterAPI
+ */
+
+/**
+ * Builds the api object that will be returned from the register api method.
+ * Also conducts certain side-effects, such as harvesting a PageView event when triggered and gathering metadata for the registered entity.
+ * @param {Object} agentRef the reference to the base agent instance
+ * @param {Object} handlers the shared handlers to be used by both the base agent's API and the external target's API
+ * @param {Object} target the target information to be used by the external target's API to send data to the correct location
+ * @returns {RegisterAPI} the api object to be returned from the register api method
+ */
+export function buildRegisterApi (agentRef, handlers, target) {
+  const attrs = {}
+  warn(53, 'newrelic.register')
+  if (!isValidTarget(target, false)) {
+    const invalidTargetResponse = () => warn(47, target)
+    invalidTargetResponse()
+    return {
+      addPageAction: invalidTargetResponse,
+      log: invalidTargetResponse,
+      noticeError: invalidTargetResponse,
+      setApplicationVersion: invalidTargetResponse,
+      setCustomAttribute: invalidTargetResponse,
+      setUserId: invalidTargetResponse,
+      /** metadata */
+      metadata: {
+        customAttributes: attrs,
+        target
+      }
+    }
+  }
+
+  const waitForRumResponse = new Promise((resolve, reject) => {
+    /** if the connect callback doesnt resolve in 15 seconds... reject */
+    const timeout = setTimeout(reject, 15000)
+    handle('api-pve', [
+      /** Handles the rum response when finished */
+      (data) => {
+        try {
+          clearTimeout(timeout)
+          const entityGuid = data.app.agents?.[0].entityGuid
+          if (!entityGuid) throw new Error('No target entity guid returned from PVE call')
+          target.entityGuid = entityGuid
+          resolve(data)
+        } catch (err) {
+          reject(err)
+        }
+      },
+      /** custom attributes to send with the rum call */
+      attrs,
+      /** target to send the rum call to (we dont have the entityguid yet in PVE call, so must supply a direct obj) */
+      target
+    ], undefined, FEATURE_NAMES.pageViewEvent, agentRef.ee)
+  })
```

Comment:
It's very jarring mixing native Promise and the EE system and buried callback for this process. I roughly see what it's doing (only because I know the rest of the code well and how it plays out) but I still don't have a precise grasp on what's happening and when. Which is to say, this is overly complicated & convoluted. Let's talk about restructuring this.

Reply:
Yeah i mentioned in a previous comment that we could probably expose some promises in the PVE chain, maybe that would help this feel more simple (by way of everything is handling async the same way)